src/connection/mod.rs

```diff
@@ -202,16 +203,101 @@ fn get(path: dbus::Path) -> Result<Connection, String> {
         }
     }
 
+    update_state(&mut connection).unwrap();
+
     Ok(connection)
 }
 
-#[derive(Default, Debug)]
+fn extract_sort_value(connection: Connection) -> i32 {
+    connection.path.rsplit('/').nth(0).unwrap().parse::<i32>().unwrap()
+}
+
+fn update_state(connection: &mut Connection) -> Result<(), String> {
+    let active_paths: Vec<String> = dbus_property!(NM_SERVICE_MANAGER,
+                                                   NM_SERVICE_PATH,
+                                                   NM_SERVICE_INTERFACE,
+                                                   "ActiveConnections")
+        .unwrap()
+        .inner::<&Vec<dbus::MessageItem>>()
+        .unwrap()
+        .iter()
+        .map(|p| dbus_path_to_string(p.inner::<&dbus::Path>().unwrap().to_owned()))
+        .collect();
+
+    let settings_paths = active_paths.iter().map(|p| {
+        dbus_path_to_string(dbus_property!(NM_SERVICE_MANAGER,
+                                           p,
+                                           NM_ACTIVE_INTERFACE,
+                                           "Connection")
+            .unwrap()
+            .inner::<&dbus::Path>()
+            .unwrap()
+            .to_owned())
+    });
+
+    connection.active_path = "".to_string();
+    connection.state = ConnectionState::Deactivated;
+
+    for (active_path, settings_path) in active_paths.iter().zip(settings_paths) {
+        if connection.path == settings_path {
+            connection.active_path = active_path.to_owned();
+            connection.state = ConnectionState::from(dbus_property!(NM_SERVICE_MANAGER,
+                                                                    connection.active_path
+                                                                        .clone(),
+                                                                    NM_ACTIVE_INTERFACE,
+                                                                    "State")
+                .unwrap_or(dbus::MessageItem::UInt32(u32::from(ConnectionState::Deactivated)))
+                .inner::<u32>()
+                .unwrap());
+            break;
+        }
+    }
+
+    Ok(())
+}
+
+fn wait(connection: &mut Connection,
```

Comment:
@josephroberts, ok. But only if both usage are very similar otherwise keep it like that.